File: src/eko-core/prompt/components/planning.ts
/**
 * 规划相关的提示词组件
 */

import { ComponentConfig, PromptComponent } from './types'

/**
 * 规划指导原则组件
 */
export const planningGuidelinesComponent: ComponentConfig = {
  id: PromptComponent.PLANNING_GUIDELINES,
  name: 'Planning Guidelines',
  description: 'Planning guidance principles',
  required: false,
  template: `# Planning Guidelines
- Adaptive planning: Single node for simple tasks, multiple nodes for complex tasks only when necessary
- Sequential execution: Break down tasks into logical sequential steps
- Tool utilization: Make use of available tools and capabilities
- Efficient planning: Focus on the most direct path to complete the user's task`,
  fn: async () => {
    return planningGuidelinesComponent.template!
  },
}

/**
 * 规划示例组件
 */
export const planningExamplesComponent: ComponentConfig = {
  id: PromptComponent.PLANNING_EXAMPLES,
  name: 'Planning Examples',
  description: 'Planning examples',
  required: false,
  template: `# Planning Examples
{examples}`,
  fn: async () => {
    // 规划示例将在后续实现
    return undefined
  },
}

/**
 * 输出格式组件
 */
export const outputFormatComponent: ComponentConfig = {
  id: PromptComponent.OUTPUT_FORMAT,
  name: 'Output Format',
  description: 'Output format description',
  required: false,
  template: `# Output Format
{format}`,
  fn: async () => {
    // 输出格式将在后续实现
    return undefined
  },
}

File: src-tauri/src/agent/prompt/components/planning.rs
use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::Arc;

use anyhow::{anyhow, Context};

use crate::agent::config::PromptComponent;
use crate::agent::error::AgentResult;
use crate::agent::prompt::components::types::{ComponentContext, ComponentDefinition};
use crate::agent::prompt::template_engine::TemplateEngine;

pub fn definitions() -> Vec<Arc<dyn ComponentDefinition>> {
    vec![
        Arc::new(PlanningGuidelinesComponent),
        Arc::new(PlanningExamplesComponent),
        Arc::new(OutputFormatComponent),
    ]
}

struct PlanningGuidelinesComponent;

#[async_trait]
impl ComponentDefinition for PlanningGuidelinesComponent {
    fn id(&self) -> PromptComponent {
        PromptComponent::PlanningGuidelines
    }

    fn name(&self) -> &str {
        "Planning Guidelines"
    }

    fn description(&self) -> &str {
        "Planning guidance principles"
    }

    fn required(&self) -> bool {
        false
    }

    fn dependencies(&self) -> &[PromptComponent] {
        &[]
    }

    fn default_template(&self) -> Option<&str> {
        Some(
            r#"# Planning Guidelines
- Adaptive planning: create the minimal number of nodes required to finish the task
- Sequential execution: break down tasks into logical steps and order them clearly
- Tool utilization: reference available tools when describing steps that require them
- Efficient planning: focus on the most direct path to complete the user's request"#,
        )
    }

    async fn render(
        &self,
        _context: &ComponentContext,
        template_override: Option<&str>,
    ) -> AgentResult<Option<String>> {
        let template = template_override
            .or_else(|| self.default_template())
            .context("missing planning guidelines template")?;

        let result = TemplateEngine::new()
            .resolve(template, &HashMap::new())
            .map_err(|e| anyhow!("failed to render planning guidelines template: {}", e))?;
        Ok(Some(result))
    }
}

struct PlanningExamplesComponent;

#[async_trait]
impl ComponentDefinition for PlanningExamplesComponent {
    fn id(&self) -> PromptComponent {
        PromptComponent::PlanningExamples
    }

    fn name(&self) -> &str {
        "Planning Examples"
    }

    fn description(&self) -> &str {
        "Planning examples"
    }

    fn required(&self) -> bool {
        false
    }

    fn dependencies(&self) -> &[PromptComponent] {
        &[]
    }

    fn default_template(&self) -> Option<&str> {
        None
    }

    async fn render(
        &self,
        _context: &ComponentContext,
        _template_override: Option<&str>,
    ) -> AgentResult<Option<String>> {
        Ok(None)
    }
}

struct OutputFormatComponent;

#[async_trait]
impl ComponentDefinition for OutputFormatComponent {
    fn id(&self) -> PromptComponent {
        PromptComponent::OutputFormat
    }

    fn name(&self) -> &str {
        "Output Format"
    }

    fn description(&self) -> &str {
        "Output format description"
    }

    fn required(&self) -> bool {
        false
    }

    fn dependencies(&self) -> &[PromptComponent] {
        &[]
    }

    fn default_template(&self) -> Option<&str> {
        None
    }

    async fn render(
        &self,
        _context: &ComponentContext,
        _template_override: Option<&str>,
    ) -> AgentResult<Option<String>> {
        Ok(None)
    }
}
